### ISSUE TITLE
Change detection is not triggered until I click the UI

### ISSUE THREAD

**Issue by adibbz** [ngrx/platform#3030]
## Minimal reproduction of the bug/regression with instructions:

Unfortunately, it would be pretty tough to recreate the issue in a stackblitz but the short version is we have a `loading` flag in our store that does not trigger change detection when updated.

In the component, we select it from the store and in our template use `{{ loading$ | ngrxPush }}` but the template never updates until I click or interact with the UI. Which leads me to believe it's a change detection issue. 

Printing out the flag in the page, we can see that it never does update, even though the store has been updated as we can see via ngrx devtools.

## Expected behavior:

Change detection is triggered and template updates with emitted value. 

## Versions of NgRx, Angular, Node, affected browser(s) and operating system(s):
Angular: `^11.0.9`
NgRx: `^11.0.1`
Node: `12.16.1`
Chrome: `90.0.4430.212`
OS: `Mac`

**Comment by timdeschryver** [comment]
Is the behavior the same with the Angular `async` pipe?

**Comment by adibbz** [comment]
Yup - that's the reason I actually tried out the component package. But it was the exact same behavior, had to click to trigger change detection. 

From what I was reading the async pipe won't trigger change detection when a new value from the store emits.

**Comment by timdeschryver** [comment]
If that's the case, I'm going to close this issue.
Previously in similar issues, the problem was that the state wasn't changed in an immutable way.
If you have a reproduction that contradicts this thought, feel free to reopen the issue.

If you need some help, feel free to join our [discord server](https://discord.com/invite/ngrx).

**Comment by twittwer** [comment]
@adibbz We had a similar issue. And switched to [@rx-angular/template](https://github.com/rx-angular/rx-angular/blob/master/libs/template/README.md) with its [PushPipe](https://github.com/rx-angular/rx-angular/blob/master/libs/template/docs/push.md) & [LetDirective](https://github.com/rx-angular/rx-angular/blob/master/libs/template/docs/api/let-directive.md).

@timdeschryver Am I right that `ngrxPush` doesn't do anything different from the `async` pipe as long as the NgZone is active in the application?
https://github.com/ngrx/platform/blob/3b565b47493a3a86681690893d89d28b25b949ac/modules/component/src/core/cd-aware/creator_render.ts#L12-L16

**Comment by timdeschryver** [comment]
@twittwer you're correct.

**Comment by adibbz** [comment]
Ah that makes sense. Thanks for the clarification guys. 

@twittwer I'll check out @rx-angular/template. Appreciate it!

**Comment by DzmVasileusky** [comment]
Had same problem which could be solved with `@rx-angular/template` only.

@timdeschryver 
As I understand the issue:
Both `async` and `ngrxPush` are just marking template for check with working Zone.
It seems NGRX doesn't causing `app.tick` itself or doing it in a wrong moment.
For example in my case I had a very popular `ngrx` app pattern:
getItemAction -> service.getItemFromAPI -> getItemSuccess -> update state through reducer with `loading: false` and item data.
And it seems that state was populated after the last app tick was called. So it was just marked for check but waited for the next tick.
To consistently reproduce an issue I have added a 2-3s delay to the http call.

I have set up all runtime checks for immutability / called in zone and it passed. So pretty sure there is some unknown bug / unrelated behaviour.
Tried to find an info about when NGRX is calling app.tick in case of effects but couldn't find it.

**Comment by timdeschryver** [comment]
@DzmVasileusky this is working as intended.
When the application is running with zone.js, the `push` pipe behaves the same way as the `async` pipe.


Ref: https://github.com/ngrx/platform/blob/master/modules/component/src/core/tick-scheduler.ts#L9

**Comment by DzmVasileusky** [comment]
@timdeschryver I know, I just wonder how does `ngrx` suppose to work with `async/push` pipe?
It seems now that in the most apps change detection is triggered by something unrelated to the ngrx. Which is looking like race condition.
I have spotted this issue in a very simple small app where is not much code which can trigger CD.
So I am trying to understand if NGRX is calling change detection itself in effects/actions/reducer or not? And if yes - then what is the flow.

**Comment by DzmVasileusky** [comment]
@DzmVasileusky so what is the best practice to handle data request case with ngrx?
All the best practice guides telling us to do like this

actions.ts:
```
export const getProperty = createAction(
    'Get Property',
    props<{ propertyId: string | number; }>()
);

export const getPropertySuccess = createAction(
    'Get Property Success',
    props<{ property: SomeType }>()
);

export const getPropertyError = createAction(
    'Get Property Error',
    props<{ error: string }>()
);
```

effects.ts
```
    getProperty$ = createEffect(() => this.actions$.pipe(
        ofType(actions.getProperty),
        switchMap((action) => this.service.getProperty(action.propertyId)
        .pipe(
            map((property: SomeType) =>  actions.getPropertySuccess({ property })),
            catchError((error) => of(actions.getPropertyError({ error })))
        ))
    ));
```

reducer.ts:
```
const featureReducer = createReducer(
    initialState,
    on(actions.init, (state) => ({ ...state, error: undefined, isLoading: false })),
    on(actions.getProperty, (state) => ({ ...state, isLoading: true })),
    on(actions.getPropertySuccess, (state, payload) => ({
        ...state, isLoading: false, property: payload.property
    })),
    on(actions.getPropertyError, (state, payload) => ({
        ...state, isLoading: false, error: payload.error
    }))
);
```

But this is not working sometimes since NGRX doesn't calling change detection itself, because Change Detection cycle which is caused by `this.service.getProperty(action.propertyId)` may already pass when reducer will update value in the state.

So how we suppose to design an ngrx architecture in this case?
Should we do like below?
```
this.loading = this.store.select(selectors.loading).pipe(tap(() => {

   this.cd.detectChanges();

}))
```

**Comment by nartc** [comment]
> I have spotted this issue in a very simple small app where is not much code which can trigger CD.

Can you share the app?

**Comment by srnec** [comment]
I noticed this issue as well in our really big app with tones of selectors, effects an reducers.
My naive explanation is that because of numerous emits a race condition happens. 
So we have our change/emit but it seems it occurs after app.tick thus markForCheck will mark given component but no other tick happens afterwards therefore it is missing one CD.
We have not nice workaround by adding observeOn(animationFrameScheduler) pipe. This way I assume it will handle the emissions in different time that won't cause the mentioned issue.

**Comment by DzmVasileusky** [comment]
@nartc Sorry, its private.
I will try to create a simple reproduction but not sure if it will work since we don't have this issue in a different app.
As @srnec mentioned above it is some kind of race condition probably

**Comment by BOBRIWE** [comment]
Hi all. @DzmVasileusky, we also have this issue, it also occurs randomly, for now I think this [example](https://stackblitz.com/edit/angular-ivy-j2xzse?file=src%2Fapp%2Fapp.component.ts) is very close to the issue. 

May it be the case that ngrx emits store change outside angular and at the end of event loop?

**Comment by nartc** [comment]
@DzmVasileusky I understand. Issues like this are hard to debug without a reproduction. Folks do things differently; a `cdr.detach()` somewhere; or `runOutsideAngular()` somewhere up the stack; AsyncPipe or PushPipe; OnPush or not etc... Without a reproduction, we're just guessing what might have happened. Personally, I've never had any race condition issues with NgRx so I can't really pinpoint what could be the case.

I know there are two moments that might have an effect on ChangeDetection, and none **should** be happening inside of NgRx:
- When an Action is dispatch; does the Action dispatch outside of Zone? At the moment the Action is dispatched, is it supposed to trigger CD? (eg: a click vs a rxjs#interval, these behave differently in an OnPush component)
- When a Selector is read; using AsyncPipe, PushPipe, or straight up `.subscribe()`. Again, the execution context is important. `AsyncPipe` and `PushPipe` seem to do `markForCheck` but does it work with `detechChanges` etc...

**Comment by DzmVasileusky** [comment]
@nartc We have found what was causing problems in our case

We have such a service which stores mixpanel distinctId. 
```
mixpanel.init(this.config.mixPanelKey, {
    loaded: (mixpanel: any) => {

        this.distinctId$.next(mixpanel.get_distinct_id());

    },
    persistence: 'localStorage'
});
```

And then we are passing this distinct id to any our http request
```
@Injectable()
export class MixpanelInterceptor implements HttpInterceptor {

    constructor (private analyticsService: AnalyticsService) {}

    intercept (req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {

        return this.mixpanelService.distinctId$.pipe(take(1), mergeMap((distinctId) => {

            // Users with the distinctId check somehow are still able to send 'undefined' as a string.
            // Actual undefined/null can't pass, so this might be from plugins that affect localStorage
            if (distinctId && distinctId.toString() !== 'undefined') {

                const authReq = req.clone({ setHeaders: { 'Mixpanel-Distinct-Id': distinctId.toString() } });

                return next.handle(authReq);

            }

            return next.handle(req);

        }));

    }

}
```

So it seems like `loaded` callback from mixpanel is somehow called outside of Zone so setting `this.distinctId$.next(mixpanel.get_distinct_id());` made it outside of zone as well.
Which made any subscription to `distinctId$` outside of Zone as well. So every http request haven't called app tick in our case because it was subscribed to this `distinctId$` in the interceptor.

So for anyone who will come here: your http requests can be called outside of Zone because you have called / used interceptor which is outside of Zone because of some third-party library like mixpanel or google maps (have seen issues with it before)

**Comment by brandonroberts** [comment]
Thanks for the follow-up @DzmVasileusky. Looks like things are working as expected from our side. Maybe we could look into a development-time check to surface this earlier and log a warning.

**Comment by nartc** [comment]
Thanks @DzmVasileusky for the follow up.
@brandonroberts maybe a check with NgZone on updating state and logs a warning of a state update happens outside of zone? Configurable with a flag?

**Comment by brandonroberts** [comment]
@nartc sounds reasonable, but yes on being configurable also. Completely Zoneless apps would be a different story

**Comment by brandonroberts** [comment]
We should open a new issue/discussion for this.

**Comment by timdeschryver** [comment]
Something like this? 👀
https://ngrx.io/guide/store/configuration/runtime-checks#strictactionwithinngzone

**Comment by brandonroberts** [comment]
Yes, exactly like that 😅. We still wouldn't have caught this though right?

**Comment by timdeschryver** [comment]
I don't know... 

The original question should be caught.
For https://github.com/ngrx/platform/issues/3030#issuecomment-1386819566 that's another thing because I suppose it isn't using ngrx/store to update the state.

I could be wrong though

**Comment by DzmVasileusky** [comment]
@timdeschryver it is, since http request was chained with an action through an effect
so request haven't called cd so state update was not re-rendered

**Comment by timdeschryver** [comment]
@DzmVasileusky Do you have a reproduction of that? 
It's not that I don't believe you, but I think if this was the case then we would've more issues with it.
